From 3860abed348e7157ce8fd6518729265c5c21ee34 Mon Sep 17 00:00:00 2001
From: Tim Griesbach <tim.griesbach@uni-bonn.de>
Date: Wed, 15 Jan 2025 15:23:29 +0100
Subject: [PATCH] Implement sleep and retry in case of busy file system

---
 src/sc_io.c | 67 +++++++++++++++++++++++++++++++++++++++++++++++++----
 1 file changed, 63 insertions(+), 4 deletions(-)

diff --git a/src/sc_io.c b/src/sc_io.c
index 4cfc7b81..904fc573 100644
--- a/src/sc_io.c
+++ b/src/sc_io.c
@@ -37,6 +37,34 @@
 #include <errno.h>
 #endif
 
+/** In case of an unexpected retrieved count 0, this macro sleeps and retries.
+ * In the Github Actions CI, we experienced transient CI failures due to the
+ * file system not performing requested I/O operations.
+ *
+ * The macro can be only callled in the case that expected count is unequal 0.
+ * It checks if count is 0 and if this is true, it sleeps and then retries the
+ * I/O operation once.
+ */
+#define SC_IO_SLEEP_AND_RETRY(func, time) do {\
+                                    retval = sc_io_error_class (sc_MPI_SUCCESS,\
+                                                                  &errcode);\
+                                    SC_CHECK_MPI (retval);\
+                                    if (*ocount == 0) {\
+                                      sc_sleep (time);\
+                                      mpiret = func (mpifile, offset,\
+                                                     (void *) ptr, count, t,\
+                                                     &mpistatus);\
+                                      if (mpiret == sc_MPI_SUCCESS) {\
+                                        retval = sc_MPI_Get_count (&mpistatus,\
+                                                                   t, ocount);\
+                                        SC_CHECK_MPI (retval);\
+                                      }\
+                                      retval = sc_io_error_class (mpiret,\
+                                                                  &errcode);\
+                                      SC_CHECK_MPI (retval);\
+                                    }} while (0)
+
+
 sc_io_sink_t       *
 sc_io_sink_new (int iotype, int iomode, int ioencode, ...)
 {
@@ -1723,7 +1751,15 @@ sc_io_read_at (sc_MPI_File mpifile, sc_MPI_Offset offset, void *ptr,
     /* working around 0 count not working for some implementations */
     mpiret = sc_MPI_Get_count (&mpistatus, t, ocount);
     SC_CHECK_MPI (mpiret);
-    return sc_MPI_SUCCESS;
+
+    /* This macro sleeps and tries the I/O operation again if the requested I/O
+     * operation was not performed by the file system.
+     * The macro SC_HACK_MPI_SLEEP is expected to be defined during
+     * configuration.
+     */
+    SC_IO_SLEEP_AND_RETRY (MPI_File_read_at, SC_HACK_MPI_SLEEP);
+
+    return errcode;
   }
   retval = sc_io_error_class (mpiret, &errcode);
   SC_CHECK_MPI (retval);
@@ -1806,7 +1842,14 @@ sc_io_read_at_all (sc_MPI_File mpifile, sc_MPI_Offset offset, void *ptr,
     mpiret = sc_MPI_Get_count (&mpistatus, t, ocount);
     SC_CHECK_MPI (mpiret);
 
-    return sc_MPI_SUCCESS;
+    /* This macro sleeps and tries the I/O operation again if the requested I/O
+     * operation was not performed by the file system.
+     * The macro SC_HACK_MPI_SLEEP is expected to be defined during
+     * configuration.
+     */
+    SC_IO_SLEEP_AND_RETRY (MPI_File_read_at_all, SC_HACK_MPI_SLEEP);
+
+    return errcode;
   }
 
   retval = sc_io_error_class (mpiret, &errcode);
@@ -2008,7 +2051,15 @@ sc_io_write_at (sc_MPI_File mpifile, sc_MPI_Offset offset,
     /* working around 0 count not working for some implementations */
     mpiret = sc_MPI_Get_count (&mpistatus, t, ocount);
     SC_CHECK_MPI (mpiret);
-    return sc_MPI_SUCCESS;
+
+    /* This macro sleeps and tries the I/O operation again if the requested I/O
+     * operation was not performed by the file system.
+     * The macro SC_HACK_MPI_SLEEP is expected to be defined during
+     * configuration.
+     */
+    SC_IO_SLEEP_AND_RETRY (MPI_File_write_at, SC_HACK_MPI_SLEEP);
+
+    return errcode;
   }
   retval = sc_io_error_class (mpiret, &errcode);
   SC_CHECK_MPI (retval);
@@ -2091,7 +2142,15 @@ sc_io_write_at_all (sc_MPI_File mpifile, sc_MPI_Offset offset,
     /* working around 0 count not working for some implementations */
     mpiret = sc_MPI_Get_count (&mpistatus, t, ocount);
     SC_CHECK_MPI (mpiret);
-    return sc_MPI_SUCCESS;
+
+    /* This macro sleeps and tries the I/O operation again if the requested I/O
+     * operation was not performed by the file system.
+     * The macro SC_HACK_MPI_SLEEP is expected to be defined during
+     * configuration.
+     */
+    SC_IO_SLEEP_AND_RETRY (MPI_File_write_at_all, SC_HACK_MPI_SLEEP);
+
+    return errcode;
   }
 
   retval = sc_io_error_class (mpiret, &errcode);
-- 
2.30.2

